File: packages/sdk-ruby/lib/pocketping/ai/gemini_provider.rb
# frozen_string_literal: true

require_relative "base"

module PocketPing
  module AI
    # Google Gemini generateContent provider.
    #
    # @example
    #   provider = PocketPing::AI::GeminiProvider.new(api_key: ENV["GEMINI_API_KEY"])
    #   client = PocketPing::Client.new(ai_provider: provider)
    class GeminiProvider < Base
      DEFAULT_MODEL = "gemini-1.5-flash"
      DEFAULT_BASE_URL = "https://generativelanguage.googleapis.com/v1beta"

      # @param api_key [String] Google AI API key (required)
      # @param model [String] Model name (default: "gemini-1.5-flash")
      # @param base_url [String] API base host (default Google endpoint, overridable for tests)
      def initialize(api_key:, model: DEFAULT_MODEL, base_url: DEFAULT_BASE_URL)
        super()
        raise ArgumentError, "api_key is required" if api_key.nil? || api_key.empty?

        @api_key = api_key
        @model = model
        @base_url = base_url.chomp("/")
      end

      # @return [String] "gemini"
      def name
        "gemini"
      end

      # @param messages [Array<PocketPing::Message>]
      # @param system_prompt [String, nil]
      # @return [String]
      def generate_response(messages, system_prompt = nil)
        contents = messages.map do |message|
          {
            role: message.sender == Sender::VISITOR ? "user" : "model",
            parts: [{ text: message.content }]
          }
        end

        # Prepend the system prompt to the first user turn, if any.
        if system_prompt && (first = contents.find { |c| c[:role] == "user" })
          original = first[:parts][0][:text]
          first[:parts][0][:text] = "#{system_prompt}\n\nUser: #{original}"
        end

        body = {
          contents: contents,
          generationConfig: {
            maxOutputTokens: 1000,
            temperature: 0.7
          }
        }

        url = "#{@base_url}/models/#{@model}:generateContent?key=#{@api_key}"
        data = post_json(url, body)
        data.dig("candidates", 0, "content", "parts", 0, "text") || ""
      end

      # Gemini has no lightweight health endpoint; available when a key is set.
      #
      # @return [Boolean]
      def available?
        !@api_key.nil? && !@api_key.empty?
      end
    end
  end
end

File: packages/sdk-php/src/AI/GeminiProvider.php
<?php

declare(strict_types=1);

namespace PocketPing\AI;

use PocketPing\Http\CurlHttpClient;
use PocketPing\Http\HttpClientInterface;
use PocketPing\Models\Message;
use PocketPing\Models\Sender;

/**
 * AI provider backed by the Google Gemini generateContent API.
 */
final class GeminiProvider implements AIProviderInterface
{
    public const DEFAULT_BASE_URL = 'https://generativelanguage.googleapis.com/v1beta';
    public const DEFAULT_MODEL = 'gemini-1.5-flash';

    private readonly string $apiKey;
    private readonly string $model;
    private readonly string $baseUrl;
    private readonly HttpClientInterface $httpClient;

    public function __construct(
        string $apiKey,
        string $model = self::DEFAULT_MODEL,
        string $baseUrl = self::DEFAULT_BASE_URL,
        ?HttpClientInterface $httpClient = null,
    ) {
        $this->apiKey = $apiKey;
        $this->model = $model;
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->httpClient = $httpClient ?? new CurlHttpClient();
    }

    public function name(): string
    {
        return 'gemini';
    }

    public function generateResponse(array $messages, ?string $systemPrompt = null): string
    {
        $contents = [];
        foreach ($messages as $msg) {
            $role = $msg->sender === Sender::VISITOR ? 'user' : 'model';
            $contents[] = [
                'role' => $role,
                'parts' => [['text' => $msg->content]],
            ];
        }

        // Prepend the system prompt to the first user message, if any.
        if ($systemPrompt !== null && $systemPrompt !== '' && $contents !== [] && $contents[0]['role'] === 'user') {
            $contents[0]['parts'][0]['text'] = "{$systemPrompt}\n\nUser: {$contents[0]['parts'][0]['text']}";
        }

        $url = "{$this->baseUrl}/models/{$this->model}:generateContent?key={$this->apiKey}";

        $response = $this->httpClient->post(
            $url,
            [
                'contents' => $contents,
                'generationConfig' => [
                    'maxOutputTokens' => 1000,
                    'temperature' => 0.7,
                ],
            ],
        );

        if ($response['error'] !== null || $response['httpCode'] < 200 || $response['httpCode'] >= 300) {
            throw new \RuntimeException(
                sprintf('Gemini API error: %s', $response['error'] ?? "HTTP {$response['httpCode']}")
            );
        }

        $data = json_decode((string) $response['body'], true);
        if (!is_array($data)) {
            return '';
        }

        return (string) ($data['candidates'][0]['content']['parts'][0]['text'] ?? '');
    }

    public function isAvailable(): bool
    {
        if ($this->apiKey === '') {
            return false;
        }

        $response = $this->httpClient->get("{$this->baseUrl}/models?key={$this->apiKey}");

        return $response['error'] === null
            && $response['httpCode'] >= 200
            && $response['httpCode'] < 300;
    }
}
